promql/engine.go

```diff
@@ -2835,16 +2835,70 @@ func vectorElemBinop(op parser.ItemType, lhs, rhs float64, hlhs, hrhs *histogram
 	case parser.MOD:
 		return math.Mod(lhs, rhs), nil, true, nil
 	case parser.EQLC:
+		if hlhs != nil && hrhs == nil {
+			return 0, nil, false, annotations.NewIncompatibleTypesInBinOpInfo("histogram", "==", "float", pos)
+		}
+		if hlhs == nil && hrhs != nil {
+			return 0, nil, false, annotations.NewIncompatibleTypesInBinOpInfo("float", "==", "histogram", pos)
+		}
+		if hlhs != nil && hrhs != nil {
+			return 0, hlhs, hlhs.Equals(hrhs), nil
```

Comment:
Yeah, I was also wondering about that. But I think we should compact them and then compare them (To avoid situations like [this](https://github.com/prometheus/prometheus/issues/13563#issuecomment-2268008818) ). And for that we can make copies of both the histograms and then we can compact them and compare them ( so that we do not mutate them ). WDYT?

Reply:
Yes, that would be the safe way. However, copying is expensive, so I would for now go with the assumption that everything is compacted already (which is why we should clearly comment it here). If that turns out to not be true, we can either fix it and make sure that compactions really happens in all cases, or in the worst case we could still go down the path of copying and compacting.

Reply:
Yeah, then we should go without compacting for now. But testing whether compaction really happens would be better IMO.